### ISSUE TITLE
Validator rejects certain valid URIs with percent-encoded characters

### ISSUE THREAD

**Issue by AaronBallman** [Microsoft/sarif-sdk#1120]
I *think* the validator may not be properly validating URIs that have UTF-8 data which is percent encoded. The file name I was testing with was:

``Te;s%ing`this~file_out #!&^#()[]bläh.c``

The generated URI was:

``file:///C:/Users/Aaron/Desktop/Te;s%25ing%60this~file_out%20%23!&%5E%23()%5B%5Dbl%C3%A4h.c``

The tool claims that it's an invalid URI, but if I replaced the `ä` with an `a` and generate the URI again, it validates fine. However, the UTF-8 code points for `ä` are `0xC3 0xA4`, so I think this URI actually is valid, at least according to my reading of the URI spec linked to in the sarif spec.

**Comment by michaelcfanning** [comment]
@lgolding, @EasyRhinoMSFT bringing this to your particular attention as an issue to look at. we may be encountering an underlying problem in .NET itself with this one.

**Comment by lgolding** [comment]
@AaronBallman @michaelcfanning @EasyRhinoMSFT 

Yes, I can verify the failure with a unit test. In fact, I just have to paste that URI into a JSON file and the syntax squiggle shows that .NET doesn't like this URI:

![image](https://user-images.githubusercontent.com/11762653/47940505-0a17bd00-dea8-11e8-9763-7f5f4ec831a5.png)

If I replace `%C3%A4` with `a` the squiggle goes away, consistent with what Aaron saw.

Now to find out why.

**Comment by lgolding** [comment]
First of all, this is not a "combining character" issue. `c3 a4` is the single composed character

U+00E4 | ä | c3 a4 | LATIN SMALL LETTER A WITH DIAERESIS

**Comment by lgolding** [comment]
Experimenting in the C# interactive window:
```
> string uriString = "file:///C:/Users/Aaron/Desktop/Te;s%25ing%60this~file_out%20%23!&%5E%23()%5B%5Dbl%C3%A4h.c";
> Uri uri = new Uri(uriString, UriKind.Absolute);
> uri.AbsoluteUri
"file:///C:/Users/Aaron/Desktop/Te;s%25ing%60this~file_out%20%23!&%5E%23()%5B%5Dbl%C3%A4h.c"
> uri.Authority
""
> uri.Host
""
> uri.AbsolutePath
"C:/Users/Aaron/Desktop/Te;s%25ing%60this~file_out%20%23!&%5E%23()%5B%5Dbl%C3%A4h.c"
> uri.Scheme
"file"
> uri.IsAbsoluteUri
true
> uri.OriginalString
"file:///C:/Users/Aaron/Desktop/Te;s%25ing%60this~file_out%20%23!&%5E%23()%5B%5Dbl%C3%A4h.c"
> uri.IsWellFormedOriginalString()
false
```

We see that everything looks fine except for `uri.IsWellFormedOriginalString()`. What's it complaining about?

**Comment by lgolding** [comment]
Also:
```
> Uri.IsWellFormedUriString(uriString, UriKind.RelativeOrAbsolute)
false
```
which explains why the rule fires, because that's the API that the rule calls.

**Comment by lgolding** [comment]
The [docs](https://docs.microsoft.com/en-us/dotnet/api/system.uri.iswellformeduristring?view=netframework-4.7.2) say:
> Beginning in .NET 4.5, strings are always considered well-formed in accordance with RFC 3986 and RFC 3987, _whether or not IRI [International Resource Identifiers] or IDN [Internationalized Domain Name] are enabled_ [my emphasis].

We build for `net461` and `netstandard2.0`, and `netstandard2.0` [is supported](https://blogs.msdn.microsoft.com/dotnet/2017/08/14/announcing-net-standard-2-0/) on `net461` and later, so those words apply to us.

So, what does RFC 3986 say?

**Comment by lgolding** [comment]
[It says]:

> Percent-encoded octets (Section 2.1) may be used within a URI to represent characters
> outside the range of the US-ASCII coded character set if this representation is allowed
> by the scheme or by the protocol element in which the URI is referenced. 
> Such a definition should specify the character encoding used to map those characters to
> octets prior to being percent-encoded for the URI.

In our case, the character is `ä`, the encoding is UTF-8, and the resulting octets are `c3 a4`. Perhaps we have to tell the URI object what encoding was used.

**Comment by lgolding** [comment]
@AaronBallman @michaelcfanning @EasyRhinoMSFT 

The problem is that the string Aaron provided is improperly escaped.

Let's start with the unescaped file name that Aaron provided:
```
Te;s%ing`this~file_out #!&^#()[]bläh.c
```
... and let's prepend `file:///C:/dir/` to it, giving
```
file:///C:/dir/Te;s%ing`this~file_out #!&^#()[]bläh.c
```

Next, let's let .NET escape the string by calling `Uri.EscapeUriString`. The result is
```
file:///C:/dir/Te;s%25ing%60this~file_out%20#!&%5E#()[]bl%C3%A4h.c
```

If I pass the resulting string to `Uri.IsWellFormedUriString` it returns true, the JSON editor does not squiggle it, and my unit test passes.

But the "escaped" string Aaron provided is (modulo the prefix):
```
file:///C:/dir/Te;s%25ing%60this~file_out%20%23!&%5E%23()%5B%5Dbl%C3%A4h.c
```

You see that this version escaped the `#`, `[`, and `]` characters, which `Uri.EscapeUriString` does not. Why this makes a difference I don't know. But the bottom line is, if your file names might have non-ASCII characters in them, then escape them by calling `Uri.EscapeUriString` before creating the `Uri` object, and all will be well.

**Comment by AaronBallman** [comment]
> You see that this version escaped the `#`, `[`, and `]` characters, which `Uri.EscapeUriString` does not. Why this makes a difference I don't know. I also don't know why replacing `ä` with `a` makes a difference. But the bottom line is, if your file names might have non-ASCII characters in them, then escape them by calling `Uri.EscapeUriString` before creating the `Uri` object, and all will be well.

I do not think this behavior matches the URI spec. For starters, the `#`, `[`, and `]` characters are all general delimiters, which are reserved. See RFC 3986 2.2:

> URI producing applications should percent-encode data octets that
   correspond to characters in the reserved set unless these characters
   are specifically allowed by the URI scheme to represent data in that
   component.  If a reserved character is found in a URI component and
   no delimiting role is known for that character, then it must be
   interpreted as representing the data octet corresponding to that
   character's encoding in US-ASCII.

I believe this means that even if those characters are not used in a delimiting role, they still should be decoded and interpreted as US-ASCII characters. Also, `#` has special meaning in path components per Section 3.3, which states that `#` terminates the path component.

As for the UTF-8 encoding, from RFC 3986 section 2.5:

> When a new URI scheme defines a component that represents textual
   data consisting of characters from the Universal Character Set [UCS],
   the data should first be encoded as octets according to the UTF-8
   character encoding [STD63]; then only those octets that do not
   correspond to characters in the unreserved set should be percent-
   encoded.  For example, the character A would be represented as "A",
   the character LATIN CAPITAL LETTER A WITH GRAVE would be represented
   as "%C3%80", and the character KATAKANA LETTER A would be represented
   as "%E3%82%A2".

However, I noticed that this issue only arises when # and UTF-8 are both encoded in the same path!

`file:///C:/dir/blah.c` validates
`file:///C:/dir/bl%C3%A4h.c` validates
`file:///C:/dir/bl%23h.c` validates
`file:///C:/dir/b%23l%C3%A4h.c` does NOT validate
`file:///C:/dir/bl%C3%A4h.c%23` does NOT validate

So I'm still not certain why this is invalid per the reference spec. Perhaps .NET references a subtly different RFC?

> If you're not using .NET to produce your SARIF, I don't know what API you should call to produce a properly escaped URI string that the .NET `Uri` class will accept. Whatever API is used, the result should be as follows:
> 
> > The generic URI syntax recommends that new URI schemes that provide for the representation of character data in a URI should, in effect, _represent characters from the unreserved set without translation, and should convert all other characters to bytes according to UTF-8, and then percent-encode those values_.  [My emphasis] This suggestion was introduced in January 2005 with the publication of RFC 3986. URI schemes introduced before this date are not affected.

That is what I've done, though.

My scheme is: the entire string is encoded as UTF-8 already; if a character from the string is allowed by the pchar production in section 3.3 (`pchar  = unreserved / pct-encoded / sub-delims / ":" / "@"`), it gets written out directly, otherwise it gets percent encoded.

> -- https://en.wikipedia.org/wiki/Percent-encoding
> 
> Note that the last sentence means that URIs that use the `http` and `file` schemes are _not required_ to behave this way. Depending on the API you use to percent encode your URI, you might or might not get a result that the .NET `Uri` class will accept. I don't know what to do about that.

That is unfortunate.

I still think there's something fishy going on here as there certainly needs to be a way to specify a URI containing UTF-8 data and a `#` ASCII character that does not terminate the path component.

**Comment by lgolding** [comment]
Ah, I think .NET uses an earlier RFC because they need to be back-compatible with versions of .NET that predate RFC 3986.

**Comment by AaronBallman** [comment]
So it sounds like it's a bug with the validator because the SARIF spec specifies RFC 3986 but .NET targets something slightly different. If that turns out to be true (maybe I'm misunderstanding the RFC), I'd say this bug can be closed as Won't Fix and it could be added to a list of known issues.

**Comment by lgolding** [comment]
I guess that's right. And -- hah! -- the SDK _does_ need a list of know issues. I'll start one. Thanks

**Comment by lgolding** [comment]
No, that's not right. See the documentation for [`System.Uri`](https://docs.microsoft.com/en-us/dotnet/api/system.uri?view=netframework-4.7.2), in particular the section International Resource Identifiers:

> The specifications for URIs are documented in RFC 2396, RFC 2732, **RFC 3986, and RFC 3987** published by the Internet Engineering Task Force (IETF).
>
> With the growth of the Internet, there is a growing need to identify resources using languages other than English. Identifiers which facilitate this need and allow non-ASCII characters (characters in the Unicode/ISO 10646 character set) are known as International Resource Identifiers (IRIs). **The specifications for IRIs are documented in RFC 3987 published by IETF. Using IRIs allows a URL to contain Unicode characters.**
>
> The existing Uri class has been extended in .NET Framework v3.5, 3.0 SP1, and 2.0 SP1 to provide IRI support based on RFC 3987. Users of .NET Framework versions before version 4.5 will not see any change from the .NET Framework 2.0 behavior unless they specifically enable IRI. This ensures application compatibility with prior versions of the .NET Framework.
> ...
> **Users of .NET Framework 4.5 and newer always have IRI enabled.**

[**Emphasis** added]

This is either a bug in the .NET `Uri` class, or -- considering how carefully the authors seem to have been about this -- something we don't understand about the spec. I've never read 3987 (only 3986), so I need to do that before drawing a conclusion.

**Comment by lgolding** [comment]
Here's the first thing we learn:

> This document defines a new protocol element, the Internationalized Resource Identifier (IRI), as a complement to the Uniform Resource Identifier (URI). An IRI is a sequence of characters from the Universal Character Set (Unicode/ISO 10646).  A mapping from IRIs to URIs is defined, which means that IRIs can be used instead of URIs, where appropriate, to identify resources.
>
> The approach of defining a new protocol element was chosen instead of extending or changing the definition of URIs.  This was done in order to allow a clear distinction and to avoid incompatibilities with existing software.  Guidelines are provided for the use and deployment of IRIs in various protocols, formats, and software components that currently deal with URIs.

Wow. TIL. So 3987 is a completely independent spec, and according to the .NET docs above, in .NET 4.5 and above the `System.Uri` class actually implements the IRI spec, rather than the URI spec.

**Comment by lgolding** [comment]
(and RFC 3987 was written by Martin Duerst and Michel Suignard of Microsoft -- so the odds are excellent that `System.Uri` implements the spec correctly.)

**Comment by AaronBallman** [comment]
I wonder if you can use `IriParsingElement` to control this (https://docs.microsoft.com/en-us/dotnet/api/system.configuration.iriparsingelement?view=netframework-4.7.2) or use `GenericUriParser` instead (https://docs.microsoft.com/en-us/dotnet/api/system.genericuriparser?view=netframework-4.7.2)?

**Comment by lgolding** [comment]
I don't think do. The docs for `System.Uri` say that IRI parsing is always on for .NET >= 4.5, but you have to turn it on for .NET < 4.5.

**Comment by AaronBallman** [comment]
I got the idea from those same docs (https://docs.microsoft.com/en-us/dotnet/api/system.uri?view=netframework-4.7.2):

> IRI and IDN processing in the Uri class can also be controlled using the System.Configuration.IriParsingElement, System.Configuration.IdnElement, and System.Configuration.UriSection configuration setting classes. The System.Configuration.IriParsingElement setting enables or disables IRI processing in the Uri class. The System.Configuration.IdnElement setting enables or disables IDN processing in the Uri class. The System.Configuration.IriParsingElement setting also indirectly controls IDN. IRI processing must be enabled for IDN processing to be possible. If IRI processing is disabled, then IDN processing will be set to the default setting where the .NET Framework 2.0 behavior is used for compatibility and IDN names are not used.

> The configuration setting for the System.Configuration.IriParsingElement and System.Configuration.IdnElement will be read once when the first System.Uri class is constructed. Changes to configuration settings after that time are ignored.

> The System.GenericUriParser class has also been extended to allow creating a customizable parser that supports IRI and IDN. The behavior of a System.GenericUriParser object is specified by passing a bitwise combination of the values available in the System.GenericUriParserOptions enumeration to the System.GenericUriParser constructor. The GenericUriParserOptions.IriParsing type indicates the parser supports the parsing rules specified in RFC 3987 for International Resource Identifiers (IRI). Whether IRI is used is dictated by the configuration values previously discussed.

Perhaps these classes don't do what's needed, however. I'm definitely not a .NET expert. :-) However, the `The System.Configuration.IriParsingElement setting enables or disables IRI processing in the Uri class.` part makes me think you can disable IRI processing for the Uri class which maybe will go back to accepting percent encoded UTF-8 data?

**Comment by lgolding** [comment]
Changed resolution to "by design". The behavior of the `System.Uri` class is correct. The URI @AaronBallman used in his example really is invalid, because it includes a fragment portion that includes the characters `#`, `[`, and `]`. These characters are invalid in a fragment, and RFC 3987 says that them MUST NOT be percent encoded. `System.Uri` correctly encodes Aaron's example, and then correctly concludes that the result is invalid.

**Comment by AaronBallman** [comment]
I'm still not on the same page, and I'm sorry if I'm being dense. The SARIF spec makes no mention of RFC 3987 and I still haven't seen rationale that the URI I provided does not conform to RFC 3986, which the SARIF spec does reference. As best I can tell, the URI is an invalid IRI but there is no requirement for it to be a valid IRI in the first place.

**Comment by lgolding** [comment]
The URI is an invalid URI because its fragment includes the characters `#`, `[`, `]`.

In email just now (you might not have seen it yet), I proposed to add the following to the spec:

> If the resource specified by a URI-valued property is actually identified by an Internationalized Resource Identifier (IRI) [RFC3987], then the SARIF producer SHALL first transform the IRI into a URI, using the mapping mechanism specified in RFC 3987 §3.1, and then assign the transformed value to the SARIF property.

**Comment by lgolding** [comment]
Also, to your statement "the URI is an invalid IRI" -- that is not correct. According to RFC 3987, _every_ valid URI is a valid IRI. It is not the case that your example is a valid URI but not a valid IRI. It is neither :-)

**Comment by lgolding** [comment]
I filed oasis-tcs/sarif-spec#275, "Specify how to store IRIs in URI-valued properties".

**Comment by AaronBallman** [comment]
> The URI is invalid because its fragment includes the characters `#`, `[`, `]`.

No, it does not. The URI in question is:

`file:///C:/Users/Aaron/Desktop/Te;s%25ing%60this~file_out%20%23!&%5E%23()%5B%5Dbl%C3%A4h.c`

This URI does not contain those ASCII characters; it contains the percent encoded value representing those ASCII characters, which is necessary to encode the path from the file system. Otherwise, there would be no way to encode a file named `foo#bar` because the percent encoded representation of `#` would be processed as a fragment that terminates the path.

**Comment by lgolding** [comment]
Sorry, I misread what you wrote at the top:

> The file name I was testing with was:
> 
> "Te;s%ing`this~file_out #!&^#()[]bläh.c"

... and I didn't notice (or at least, in the course of this long discussion I forgot) that the actual URI it was complaining about was
```
file:///C:/Users/Aaron/Desktop/Te;s%25ing%60this~file_out%20%23!&%5E%23()%5B%5Dbl%C3%A4h.c
```
I don't see anything wrong with that URI. I guess I need to investigate further.

**Comment by lgolding** [comment]
... which I suspect at this point means debugging into the .NET source. Oh well.

**Comment by lgolding** [comment]
1. The `file` URI has its own RFC ([RFC 8089, "The "file" URI scheme"](https://www.ietf.org/rfc/rfc8089.txt)). And its grammar does not include `query` or `fragment` constructs. That whole string `Te;s%ing`this~file_out #!&^#()[]bläh.c` really is the file name.
2. That is a valid file name (to my surprise), so the `Uri` class is _not_ rejecting your URI because of some special knowledge it has about the NTFS file system:

```
d:\src\sarif-sdk>copy .gitignore "Te;s%ing`this~file_out #!&^#()[]bläh.c"
        1 file(s) copied.

d:\src\sarif-sdk>dir Te*
 Volume in drive D has no label.
 Volume Serial Number is FEF6-8614

 Directory of d:\src\sarif-sdk

10/22/2018  09:24 AM             3,531 Te;s%ing`this~file_out #!&^#()[]bläh.c
               1 File(s)          3,531 bytes
               0 Dir(s)  1,729,082,880,000 bytes free
```
3. RFC 8089 ultimately relies on the `path-absolute` construct from RFC 3986, which comes down to the definition of  `pchar`, including `pct-encoded`. So it is _not_ the case that RFC 8089 introduces some new restriction that makes your `file` URI invalid.

Bottom line, I'm back to having no idea what's wrong, it might be a bug in `System.Uri` after all, and I'm afraid I can't spend more time on this right now.

But @AaronBallman, when you're back from your WG21 meetings, let's touch base again. I really do want to run this to ground.

**Comment by AaronBallman** [comment]
I had the chance to read through the various specs involved (RFCs 8089, 3986, and 3987) to see if there were other interactions I had missed, and I don't see any. As best I can tell, the SARIF spec defers to RFC8089, which leans heavily on RFC3986 and what I've produced should be conforming to those two specs. I don't see a path that gets me from SARIF to RFC3987 in terms of specification.

In short, I agree with all three of your points above. We can touch base again when you have some more cycles to spare on this. I appreciate all of the attention you've been giving it!

**Comment by AaronBallman** [comment]
Btw, I'm back from WG21 in case you want to touch base on this again. I think perhaps this bug should be reopened so we don't lose track of it.

**Comment by lgolding** [comment]
Agreed, reopened.
